pkg/operator/etcdcertsigner/etcdcertsignercontroller.go

```diff
@@ -178,24 +194,79 @@ func (c *EtcdCertSignerController) syncAllMasters(recorder events.Recorder) erro
 	return utilerrors.NewAggregate(errs)
 }
 
-func (c *EtcdCertSignerController) createSecretForNode(node *corev1.Node, recorder events.Recorder) error {
+// validateCertificateSAN makes sure that a given certificate is valid
+// at least for the SANs defined in the configuration.
+func verifyCertificateSAN(certData map[string][]byte, certName string, ipAddresses []string) error {
+	certificates, err := cert.ParseCertsPEM(certData["tls.crt"])
+	if err != nil {
+		return fmt.Errorf("could not parse TLS cert for %s: %w", certName, err)
+	}
+	// Using the first certificate to verify the IP addresses
+	certificate := certificates[0]
+	for _, ipAddress := range ipAddresses {
+		if err := certificate.VerifyHostname(ipAddress); err != nil {
+			return fmt.Errorf("SAN for the certificate %s does not include %s: %w", certName, ipAddress, err)
+		}
+	}
+	return nil
+}
+
+func (c *EtcdCertSignerController) checkSecrets(node *corev1.Node, nodeInternalIPs []string, recorder events.Recorder) error {
+	etcdPeerClientCertName := tlshelpers.GetPeerClientSecretNameForNode(node.Name)
+	etcdServingCertName := tlshelpers.GetServingSecretNameForNode(node.Name)
+	metricsServingCertName := tlshelpers.GetServingMetricsSecretNameForNode(node.Name)
+
+	certNames := []string{etcdPeerClientCertName, etcdServingCertName, metricsServingCertName}
+
+	errs := []error{}
+	for _, certName := range certNames {
+		secret, err := c.secretLister.Secrets(operatorclient.TargetNamespace).Get(etcdPeerClientCertName)
+		if err != nil {
+			errs = append(errs, err)
+			continue
+		}
+		if secret.Data == nil {
+			errs = append(errs, fmt.Errorf("secret data not found in %s", certName))
+			continue
+		}
+	}
+	if len(errs) > 0 {
+		return utilerrors.NewAggregate(errs)
+	}
+
+	return nil
+}
+
+// Validates IP addresses in the SAN fields of the certificates for a given node
+func (c *EtcdCertSignerController) validateIpaddrs(node *corev1.Node, nodeInternalIPs []string, recorder events.Recorder) error {
 	etcdPeerClientCertName := tlshelpers.GetPeerClientSecretNameForNode(node.Name)
 	etcdServingCertName := tlshelpers.GetServingSecretNameForNode(node.Name)
 	metricsServingCertName := tlshelpers.GetServingMetricsSecretNameForNode(node.Name)
 
-	var err error
-	_, err = c.secretLister.Secrets(operatorclient.TargetNamespace).Get(etcdPeerClientCertName)
-	peerClientCertOk := err == nil
-	_, err = c.secretLister.Secrets(operatorclient.TargetNamespace).Get(etcdServingCertName)
-	servingCertOk := err == nil
-	_, err = c.secretLister.Secrets(operatorclient.TargetNamespace).Get(metricsServingCertName)
-	metricsCertOk := err == nil
+	certNames := []string{etcdPeerClientCertName, etcdServingCertName, metricsServingCertName}
 
-	// if we have all the certs we want, do nothing.
-	if peerClientCertOk && servingCertOk && metricsCertOk {
-		return nil
+	errs := []error{}
+	for _, certName := range certNames {
+		secret, err := c.secretLister.Secrets(operatorclient.TargetNamespace).Get(etcdPeerClientCertName)
```

Comment:
Should the secret being retrieved be `certName` rather than `etcdPeerClientCertName`?